### ISSUE TITLE
Idea: allow for an “Alt" key to make 3 or 4 layers of switch values

### ISSUE THREAD

**Issue by jasonbirchler** [markwheeler/midi-mixer-one#4]
hi 👋

I’m using midi-mixer-one with the 1010 Bluebox. I have 8 mixer channels assigned to the 8 columns of pots. I’d also like to be able to use 8 of the switches in 4 different “layers”. Allowing for one of the switches to act as an “Alt” key, in addition to the Shift key, would mean that 8 of the 10 remaining keys could easy map to each column of pots for things like Mute/Solo/Rec.

For example: key1 sends a CC that arms track 1 for rec, shift+key1 sends a CC that mutes track 1, alt+key1 sends a CC that solos track 1. Alt+Shift+key1 could also allow for a 4th layer and 40 total CC’s that could be sent via the switches. Or the Alt+Shift combo could be the “Send All CCs” command, for a total of 30 CC’s + Send All CC's

I forked the repo and started to code this, but quickly ran into an out of bounds error when trying to add the extra arrays for ```switchAltValues``` and ```switchAltShiftValues``` in ```serialize()``` and realized that I a) may have added more data than can be serialized, and b) don’t know jack squat about serializing MIDI data 😁

so before I go down a dark rabbit hole of trying to understand how MIDI data is sent over the wire, does this even sound possible? If so, does it sound like a worthwhile addition to the firmware? If so, I’m happy to contribute a PR adding the logic in the arduino code and necessary forms on the web-config tool. However, I may need some guidance on the un/serialize piece.

jason

**Comment by markwheeler** [comment]
The serialization is really just to work around the limited EEPROM storage on the Teensy LC (1024 bits). It's not particularly elegant code but it means we can store as much as possible in that space by reducing all the values to the minimum number of bits required and then writing them in byte-size chunks still using the `EEPROM.read/update` methods. TBH it might have been smarter to just create my own versions of those methods that do what we need but that's the solution I went with at the time :)

If you need to store significantly more data then it likely won't fit in the EEPROM but I think what you're describing would fit fine.

I'd suggest getting familiar with the JS code first because it's a bit more readable IMO. Take a look here to get an idea of the data structure: https://github.com/markwheeler/midi-mixer-one/blob/9becc82a8cd30aa25db95caa63e655ebd8a8d51b/web-config/js/Device.js#L54

If you get it all working, I'd be happy to have it in this repo!

**Comment by jasonbirchler** [comment]
Thanks, that helps! I didn’t even think of looking at the JS to understand the data structure, and JS is what I’m most familiar with.

For my understanding, all of the de/serialization is to support programming via the web config, correct?

**Comment by markwheeler** [comment]
yes, if you don't need configuration you could just hard code all the MIDI values and ignore the EEPROM.
